### PR TITLE
Describe implied bounds

### LINKED ISSUE #369: Describe implied bounds
https://rust-lang.github.io/rfcs/2093-infer-outlives.html

### PR DESCRIPTION
**Pull request by kirtchev-adacore**
Fixes #369 .

### CODE REVIEW

File: src/types-and-traits.rst [hunk 1]
```diff
@@ -2637,6 +2637,30 @@ A :t:`higher-ranked trait bound` is a :t:`bound` that specifies an infinite
...
+:t:`[abstract data type]s` or :t:`[function]s`.
+
+:dp:`fls_IfHRxSasGAih`
+A :t:`reference` of the form ``&'a T``, where ``'a`` is a
+:t:`lifetime parameter` and ``T`` is a :t:`type parameter`, yields
+:t:`implied bound` ``T: 'a``.
+
+:dp:`fls_SMbHOyJgVeKB`
+A :t:`reference` of the form ``&'a &'b``, where ``'a`` and ``'b`` are
+:t:`[lifetime parameter]s``, yields :t:`implied bound` ``'a: 'b``.
```
Comment (jonas-schievink) [inline src/types-and-traits.rst:2652]: I think these are swapped – but then, this is equivalent to the previous rule, so it seems redundant

File: src/types-and-traits.rst [hunk 2]
```diff
@@ -2637,6 +2637,30 @@ A :t:`higher-ranked trait bound` is a :t:`bound` that specifies an infinite
...
+
+:dp:`fls_K8nPGP5xbLb7`
+If an :t:`implied bound` applies to a :t:`field`, then the :t:`implied bound`
+also applies to the :t:`abstract data type` where the :t:`field` resides.
```
Comment (jonas-schievink) [inline src/types-and-traits.rst:2656]: Bounds cannot apply to fields, so maybe this should be reworded?

File: src/types-and-traits.rst (hunk 1 above)
Comment (kirtchev-adacore) [reply src/types-and-traits.rst:2652]: Good catch. Anyway, I will merge the two rules.

File: src/types-and-traits.rst (hunk 2 above)
Comment (kirtchev-adacore) [reply src/types-and-traits.rst:2656]: It should be the type of a field. Fixed.

File: src/types-and-traits.rst [hunk 3]
```diff
@@ -2637,6 +2637,28 @@ A :t:`higher-ranked trait bound` is a :t:`bound` that specifies an infinite
...
+:dp:`fls_J9DEsd06Ttu9`
+An :t:`implied bound` is a :t:`bound` that is not expressed in syntax, but is
+is the byproduct of relations between the :t:`[generic parameter]s` of
+:t:`[abstract data type]s` or :t:`[function]s`.
+
+:dp:`fls_IfHRxSasGAih`
+A :t:`reference` of the form ``&'a generic_parameter``, where ``'a`` is a
+:t:`lifetime parameter` and ``generic_parameter`` is an arbitrary
+:t:`generic parameter`, yields :t:`implied bound``
+``generic_parameter: 'a``.
```
Comment (Veykril) [inline src/types-and-traits.rst:2649]: The generic parameter here can be any type which can be relevant if its `Foo<T>` where `T` is a generic parameter, as the lifetime of `Foo` then depends on the generic parameter `T`. that is `&'a Foo<T>` still implies `Foo<T>: 'a`

File: src/types-and-traits.rst [hunk 4]
```diff
@@ -2637,6 +2637,28 @@ A :t:`higher-ranked trait bound` is a :t:`bound` that specifies an infinite
 list of :t:`[bound]s` for all possible :t:`[lifetime]s` specified by the
 :s:`ForGenericParameterList`.
 
+:dp:`fls_J9DEsd06Ttu9`
+An :t:`implied bound` is a :t:`bound` that is not expressed in syntax, but is
+is the byproduct of relations between the :t:`[generic parameter]s` of
```
Comment (Veykril) [inline src/types-and-traits.rst:2642]: between `lifetime parameters` and parameters and return types for functions 
between `lifetime parameters`  and fields for abstract data types

File: src/types-and-traits.rst [hunk 5]
```diff
@@ -2637,6 +2637,28 @@ A :t:`higher-ranked trait bound` is a :t:`bound` that specifies an infinite
...
+:dp:`fls_IfHRxSasGAih`
+A :t:`reference` of the form ``&'a generic_parameter``, where ``'a`` is a
+:t:`lifetime parameter` and ``generic_parameter`` is an arbitrary
+:t:`generic parameter`, yields :t:`implied bound``
+``generic_parameter: 'a``.
+
+:dp:`fls_K8nPGP5xbLb7`
+If an :t:`implied bound` applies to the :t:`type` of a :t:`field`, then the
+:t:`implied bound` also applies to the :t:`abstract data type` where the
+:t:`field` resides.
```
Comment (Veykril) [inline src/types-and-traits.rst:2654]: The type of a field can also have an explicit lifetime bound which in turn will become an implied lifetime bound for the ADT here, that is outlives requirement on  the field type, be it explicit or implied are added as an implied bound to the containing ADT

File: src/types-and-traits.rst [hunk 6]
```diff
@@ -2637,6 +2637,28 @@ A :t:`higher-ranked trait bound` is a :t:`bound` that specifies an infinite
...
+
+:dp:`fls_Uw6y5ZlaK6RY`
+If an :t:`implied bound` applies to an :t:`abstract data type`, then the
+:t:`implied bound` also applies to all :t:`[function]s` that use the
+:t:`abstract data type` and to all :t:`[path]s` that resolve to the
+:t:`abstract data type`.
```
Comment (Veykril) [inline src/types-and-traits.rst:2660]: This seems poorly worded (again, implied bounds don't propagate per se, they arise from outlives bounds (`T: 'a` in general, whether they were introduced as an implication or an explicit one). This should be restarted that implied bounds are introduced when parameters or return types of functions make use of lifetime parameters declared on the function, likewise for ADTs and fields

**Review by Veykril — CHANGES_REQUESTED** [review]

File: src/types-and-traits.rst (hunk 5 above)
Comment (kirtchev-adacore) [reply src/types-and-traits.rst:2654]: Fixed.

File: src/types-and-traits.rst (hunk 6 above)
Comment (kirtchev-adacore) [reply src/types-and-traits.rst:2660]: Fixed.

File: src/types-and-traits.rst (hunk 3 above)
Comment (kirtchev-adacore) [reply src/types-and-traits.rst]: Fixed.

File: src/types-and-traits.rst [hunk 7]
```diff
@@ -2693,6 +2693,30 @@ A :t:`higher-ranked trait bound` is a :t:`bound` that specifies an infinite
...
 :s:`ForGenericParameterList`.
 
+:dp:`fls_J9DEsd06Ttu9`
+An :t:`implied bound` is a :t:`bound` that is not expressed in syntax, but is
+is the byproduct of relations between :t:`[lifetime parameter]s` and
+:t:`[function parameter]s`, between :t:`[lifetime parameter]s` and a
+:t:`return type`, and between :t:`[lifetime parameter]s` and :t:`[field]s`.
+
+is the byproduct of relations between the :t:`[generic parameter]s` of
+:t:`[abstract data type]s` or :t:`[function]s`.
```
Comment (Veykril) [inline src/types-and-traits.rst:2703]: leftover text

File: src/types-and-traits.rst [hunk 8]
```diff
@@ -2693,6 +2693,30 @@ A :t:`higher-ranked trait bound` is a :t:`bound` that specifies an infinite
...
+
+:dp:`fls_IfHRxSasGAih`
+A :t:`reference` of the form ``&'lifetime Type``, where ``'lifetime`` is a
+:t:`lifetime parameter` and ``Type`` is a :t:`type`, yields :t:`implied bound``
+``Type: 'lifetime``.
+
+:dp:`fls_K8nPGP5xbLb7`
+If an explicit :t:`bound` or an :t:`implied bound` applies to the :t:`type` of
+a :t:`field`, then this abound also applies to the related
+:t:`abstract data type` as an :t:`implied bound`.
```
Comment (Veykril) [inline src/types-and-traits.rst:2713]: only outlives bounds matter, that is bounds where the right side has lifetimes (opposed to traits), other bounds are never implied (so `T: 'a` and `'a: 'b` bounds)

File: src/types-and-traits.rst [hunk 9]
```diff
@@ -2693,6 +2693,30 @@ A :t:`higher-ranked trait bound` is a :t:`bound` that specifies an infinite
...
+
+:dp:`fls_Uw6y5ZlaK6RY`
+If an explicit :t:`bound` or an :t:`implied bound` applies to the :t:`type` of
+a :t:`function parameter` or to a :t:`return type`, then this bound also
+applies to the related :t:`function` as an :t:`implied bound`.
```
Comment (Veykril) [inline src/types-and-traits.rst:2718]: Same thing here, only outlives bounds

File: src/types-and-traits.rst (hunk 7 above)
Comment (kirtchev-adacore) [reply src/types-and-traits.rst:2703]: Fixed.

File: src/types-and-traits.rst (hunk 8 above)
Comment (kirtchev-adacore) [reply src/types-and-traits.rst]: Fixed.

File: src/types-and-traits.rst (hunk 9 above)
Comment (kirtchev-adacore) [reply src/types-and-traits.rst]: Fixed.

File: src/types-and-traits.rst [hunk 10]
```diff
@@ -2694,15 +2694,40 @@ list of :t:`[bound]s` for all possible :t:`[lifetime]s` specified by the
 :s:`ForGenericParameterList`.
 
 :dp:`fls_AzuZmR9DXSQh`
-A :t:`trait bound` with :s:`Punctuation` ``?`` is an :t:`opt-out trait bound`.
+An :t:`opt-out trait bound` is a :t:`trait bound` with :s:`Punctuation` ``?``
+that nullifies an implicitly added :t:`trait bound`.
+
+:dp:`fls_1Sm2Yq1Ow76f`
+An :t:`outlives bound` is a :t:`trait bound` which requires that a
+:t:`generic parameter` outlives a :t:`lifetime parameter`.
```
Comment (Veykril) [inline src/types-and-traits.rst:2702]:
```suggestion
:t:`lifetime parameter` or :t:`type` outlives a :t:`lifetime parameter`.
```

**Review by Veykril — APPROVED** [review]

**Comment by Veykril** [comment]
bors merge

**Pull request merged**